// func/dist/errors/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.errorTokenTypes = exports.errorScopes = exports.errorTypes = exports.errorLevels = exports.F_EFFECT = exports.F_RUNTIME_PRINT = exports.F_RUNTIME = exports.F_SYSTEM = void 0;
exports.F_SYSTEM = {
    DUPLICATE_HANDLER: 'F_SYSTEM_DUPLICATE_HANDLER',
    INVALID_PARAM_TYPE: 'F_SYSTEM_INVALID_PARAM_TYPE',
    INVALID_PARAM_VALUE: 'F_SYSTEM_INVALID_PARAM_VALUE',
    MISSING_PROVIDER: 'F_SYSTEM_MISSING_PROVIDER',
    MISSING_PARAM_TYPES: 'F_SYSTEM_MISSING_PARAM_TYPES',
    MISSING_REQUIRED_PARAM: 'F_SYSTEM_MISSING_REQUIRED_PARAM',
    UNKNOWN_HANDLER: 'F_SYSTEM_UNKNOWN_HANDLER',
    UNSUPPORTED_ARRAY_TYPE: 'F_SYSTEM_UNSUPPORTED_ARRAY_TYPE',
};
exports.F_RUNTIME = {
    HANDLER_ERROR: 'F_RUNTIME_HANDLER_ERROR',
};
exports.F_RUNTIME_PRINT = {
    MULTIPLE_OPTIONS: 'F_RUNTIME_PRINT_MULTIPLE_OPTIONS',
    PARSE: 'F_RUNTIME_PRINT_PARSE',
    UNKNOWN_OPTION: 'F_RUNTIME_PRINT_UNKNOWN_OPTION',
};
exports.F_EFFECT = {
    DEPRECATED_API: 'F_EFFECT_DEPRECATED_API',
};
var errorLevels;
(function (errorLevels) {
    errorLevels["EFFECT"] = "effect";
    errorLevels["RUNTIME"] = "runtime";
    errorLevels["RUNTIME_PRINT"] = "runtime-print";
    errorLevels["SYSTEM"] = "system";
})(errorLevels = exports.errorLevels || (exports.errorLevels = {}));
var errorTypes;
(function (errorTypes) {
    errorTypes["DEFINITION"] = "definition";
    errorTypes["DEPRECATION"] = "deprecation";
    errorTypes["HANDLER"] = "handler";
    errorTypes["INJECTION"] = "injection";
    errorTypes["INPUT"] = "input";
    errorTypes["REGISTRATION"] = "registration";
})(errorTypes = exports.errorTypes || (exports.errorTypes = {}));
var errorScopes;
(function (errorScopes) {
    errorScopes["COMMAND"] = "command";
    errorScopes["OPTION"] = "option";
})(errorScopes = exports.errorScopes || (exports.errorScopes = {}));
var errorTokenTypes;
(function (errorTokenTypes) {
    errorTokenTypes["COMMAND_ALIAS"] = "command alias";
    errorTokenTypes["COMMAND_NAME"] = "command name";
    errorTokenTypes["OPTION_ALIAS"] = "option alias";
    errorTokenTypes["OPTION_NAME"] = "option name";
})(errorTokenTypes = exports.errorTokenTypes || (exports.errorTokenTypes = {}));
