// func/dist/errors/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleEffect = exports.handleRuntimePrintError = exports.handleSystemError = exports.normalizeRuntimeError = exports.createRuntimePrintError = exports.createRuntimeError = exports.createSystemError = exports.isFuncError = exports.FuncError = void 0;
const constants_1 = require("./constants");
class FuncError extends Error {
    cause;
    code;
    details;
    level;
    type;
    constructor(params) {
        super(params.message);
        this.name = 'FuncError';
        this.code = params.code;
        this.details = params.details || {};
        this.level = params.level;
        this.type = params.type;
        this.cause = params.cause;
    }
}
exports.FuncError = FuncError;
const isFuncError = (error) => error instanceof FuncError;
exports.isFuncError = isFuncError;
const createSystemError = (code, type, message, details = {}, cause) => new FuncError({
    code,
    details,
    level: constants_1.errorLevels.SYSTEM,
    message,
    type,
    cause,
});
exports.createSystemError = createSystemError;
const createRuntimeError = (code, type, message, details = {}, cause) => new FuncError({
    code,
    details,
    level: constants_1.errorLevels.RUNTIME,
    message,
    type,
    cause,
});
exports.createRuntimeError = createRuntimeError;
const createRuntimePrintError = (code, type, message, details = {}, cause) => new FuncError({
    code,
    details,
    level: constants_1.errorLevels.RUNTIME_PRINT,
    message,
    type,
    cause,
});
exports.createRuntimePrintError = createRuntimePrintError;
const normalizeRuntimeError = (error) => {
    if ((0, exports.isFuncError)(error))
        return error;
    const nativeError = error instanceof Error ? error : new Error(String(error));
    return (0, exports.createRuntimeError)(constants_1.F_RUNTIME.HANDLER_ERROR, constants_1.errorTypes.HANDLER, nativeError.message, { error: nativeError }, nativeError);
};
exports.normalizeRuntimeError = normalizeRuntimeError;
const handleSystemError = (error) => {
    throw error;
};
exports.handleSystemError = handleSystemError;
const handleRuntimePrintError = (error, printPrevented) => {
    if (printPrevented)
        return;
    console.error(error.message);
};
exports.handleRuntimePrintError = handleRuntimePrintError;
const handleEffect = (code, type, message, details = {}) => {
    const error = new FuncError({
        code,
        details,
        level: constants_1.errorLevels.EFFECT,
        message,
        type,
    });
    console.warn(error.message);
    return error;
};
exports.handleEffect = handleEffect;
