// func/dist/src/utils/filter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeHyphen = exports.optionsToKeyValue = exports.commandsToDatas = void 0;
const metadata_1 = require("./metadata");
const commandsToDatas = (commands = []) => {
    return commands.map(fn => Object.assign({}, Reflect.getMetadata(metadata_1.metadata.COMMAND_IDENTIFIER, fn), {
        fieldOptions: (Reflect.getMetadata(metadata_1.metadata.FIELD_OPTION_IDENTIFIER, fn) || []).map((item) => Object.assign({}, item)),
        handlers: (Reflect.getMetadata(metadata_1.metadata.METHOD_HANDLER_IDENTIFIER, fn) || []).map((item) => Object.assign({}, item)),
        subOptions: (Reflect.getMetadata(metadata_1.metadata.SUB_OPTION_IDENTIFIER, fn) || []).map((item) => Object.assign({}, item)),
    }));
};
exports.commandsToDatas = commandsToDatas;
const optionsToKeyValue = (params = []) => {
    if (!params || !params.length)
        return {};
    return params.reduce((pre, current) => {
        const name = `--${current.name}`;
        const alias = current.alias ? { [`-${current.alias}`]: name } : {};
        return Object.assign({}, pre, {
            [name]: current.type,
        }, alias);
    }, {});
};
exports.optionsToKeyValue = optionsToKeyValue;
const removeHyphen = (key) => {
    return key.replace(/^[-]+/, '');
};
exports.removeHyphen = removeHyphen;
