// func/dist/src/utils/metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handlers = exports.metadata = void 0;
var metadata;
(function (metadata) {
    metadata["DESIGN_PARAM_TYPES"] = "design:paramtypes";
    metadata["DESIGN_TYPE"] = "design:type";
    metadata["HOST_PARAM_TYPES"] = "host:paramtypes";
    metadata["COMMAND_IDENTIFIER"] = "command_ident";
    metadata["FIELD_OPTION_IDENTIFIER"] = "field_option_ident";
    metadata["FIELD_CONSTRAINT_IDENTIFIER"] = "field_constraint_ident";
    metadata["METHOD_CATCH_IDENTIFIER"] = "method_catch_ident";
    metadata["METHOD_HANDLER_IDENTIFIER"] = "method_handler_ident";
    metadata["REQUIRED_FIELD_IDENTIFIER"] = "required_field_ident";
    metadata["SUB_OPTION_IDENTIFIER"] = "sub_o_ident";
    metadata["VALUE_VALIDATOR_IDENTIFIER"] = "value_validator_ident";
    metadata["FUNC_MODULE_IDENTIFIER"] = "func_module_ident";
    metadata["PARAM_INJECT_TOKEN_IDENTIFIER"] = "param_inject_token_ident";
    metadata["SERVICE_IDENTIFIER"] = "service_ident";
    metadata["HANDLER_IDENTIFIER"] = "handler_ident";
    metadata["INJECT_TOKEN_IDENTIFIER"] = "inject_t_ident";
    metadata["INJECT_INDEX_IDENTIFIER"] = "inject_i_ident";
})(metadata || (exports.metadata = metadata = {}));
var handlers;
(function (handlers) {
    handlers["COMMAND"] = "command";
    handlers["MISSING"] = "missing";
    handlers["MAJOR"] = "major";
    handlers["ERROR"] = "error";
})(handlers || (exports.handlers = handlers = {}));
