// func/dist/constants/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.errorTokenTypes = exports.errorScopes = exports.errorCodes = void 0;
var errorCodes;
(function (errorCodes) {
    errorCodes["DUPLICATE_HANDLER"] = "FUNC_DUPLICATE_HANDLER";
    errorCodes["MISSING_PARAM_TYPES"] = "FUNC_MISSING_PARAM_TYPES";
    errorCodes["MULTIPLE_OPTIONS"] = "FUNC_MULTIPLE_OPTIONS";
    errorCodes["PARSE"] = "FUNC_PARSE_ERROR";
    errorCodes["UNKNOWN_HANDLER"] = "FUNC_UNKNOWN_HANDLER";
    errorCodes["UNSUPPORTED_ARRAY_TYPE"] = "FUNC_UNSUPPORTED_ARRAY_TYPE";
})(errorCodes = exports.errorCodes || (exports.errorCodes = {}));
var errorScopes;
(function (errorScopes) {
    errorScopes["COMMAND"] = "command";
    errorScopes["OPTION"] = "option";
})(errorScopes = exports.errorScopes || (exports.errorScopes = {}));
var errorTokenTypes;
(function (errorTokenTypes) {
    errorTokenTypes["COMMAND_ALIAS"] = "command alias";
    errorTokenTypes["COMMAND_NAME"] = "command name";
    errorTokenTypes["OPTION_ALIAS"] = "option alias";
    errorTokenTypes["OPTION_NAME"] = "option name";
})(errorTokenTypes = exports.errorTokenTypes || (exports.errorTokenTypes = {}));
