// func/dist/containers/container.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Container = void 0;
const arg_1 = __importDefault(require("arg"));
const factory_1 = require("../utils/factory");
const filter = __importStar(require("../utils/filter"));
const metadata_1 = require("../utils/metadata");
const errors_1 = require("../errors");
const context_1 = require("../context");
const constraint_validator_1 = require("../options/constraint-validator");
const handler_registry_1 = require("./handler-registry");
const registration_validator_1 = require("./registration-validator");
class Container {
    argv;
    registry;
    services;
    datas;
    constructor(params, options = {}) {
        this.argv = options.argv || process.argv.slice(2);
        this.registry = new handler_registry_1.HandlerRegistry(params);
        this.services = options.services || [];
        this.datas = this.registry.datas;
        new registration_validator_1.RegistrationValidator(this.registry).validate();
    }
    async run() {
        try {
            await this.dispatch();
        }
        catch (error) {
            this.dispatchError(error);
        }
    }
    async dispatch() {
        const scope = this.scope();
        if (!scope)
            return;
        if (scope.kind === metadata_1.handlers.MISSING &&
            !this.methodHandlers(scope.handler).length) {
            this.dispatchLegacyMissing(scope.handler);
            return;
        }
        const spec = this.optionSpec(scope.handler);
        const args = this.parse(spec);
        this.throwUnknownOptions(args, scope.kind === metadata_1.handlers.COMMAND);
        const nativeOption = this.nativeOption(spec, args);
        const commandData = this.commandData(scope.handler);
        const baseInputs = scope.kind === metadata_1.handlers.COMMAND ? args._.slice(1) : args._;
        const selectedHandler = this.selectHandler(scope.handler, args, baseInputs);
        const factory = new factory_1.Factory({
            nativeOption,
            args,
            command: commandData,
            commands: this.registry.commands(),
            handler: selectedHandler.data,
            inputs: selectedHandler.inputs,
            path: selectedHandler.path,
            services: this.services,
        });
        const constructorParams = factory.getServiceParams(scope.handler);
        const instance = new scope.handler(...constructorParams);
        try {
            this.injectFields(scope.handler, instance, args, nativeOption);
            this.validateFieldValues(scope.handler, nativeOption);
            new constraint_validator_1.ConstraintValidator(this.fieldOptions(scope.handler), args, nativeOption).validate();
            const methodParams = factory.getMethodServiceParams(instance, selectedHandler.data.methodName);
            await Promise.resolve(instance[selectedHandler.data.methodName](...methodParams));
        }
        catch (error) {
            const caught = await this.dispatchLocalCatch(scope.handler, instance, error, factory);
            if (caught)
                return;
            throw error;
        }
    }
    scope() {
        const command = this.findCommand();
        if (command)
            return { handler: command, kind: metadata_1.handlers.COMMAND };
        if (this.argv[0] && !this.argv[0].startsWith('-')) {
            const missing = this.registry.missing()[0];
            if (missing)
                return { handler: missing, kind: metadata_1.handlers.MISSING };
            return undefined;
        }
        const major = this.registry.major();
        if (major)
            return { handler: major, kind: metadata_1.handlers.MAJOR };
        if (this.argv.some(item => item.startsWith('-') && item !== '-')) {
            throw (0, errors_1.createRuntimePrintError)(errors_1.F_RUNTIME_PRINT.UNKNOWN_OPTION, errors_1.errorTypes.INPUT, `Unknown option${this.argv.length > 1 ? 's' : ''}: ${this.argv.join(', ')}.`, { options: this.argv });
        }
        return undefined;
    }
    dispatchLegacyMissing(missing) {
        const args = { _: this.argv };
        const factory = new factory_1.Factory({
            nativeOption: {},
            args,
            commands: this.registry.commands(),
            inputs: this.argv,
            services: this.services,
        });
        const params = factory.getServiceParams(missing);
        new missing(...params);
    }
    parse(spec) {
        try {
            return (0, arg_1.default)(spec, { argv: this.argv, permissive: true });
        }
        catch (error) {
            throw (0, errors_1.createRuntimePrintError)(errors_1.F_RUNTIME_PRINT.PARSE, errors_1.errorTypes.INPUT, error instanceof Error ? error.message : String(error), { error }, error instanceof Error ? error : undefined);
        }
    }
    nativeOption(spec, args) {
        return Object.keys(spec).reduce((pre, key) => {
            if (!key.startsWith('--'))
                return pre;
            return Object.assign({}, pre, {
                [filter.removeHyphen(key)]: Object.prototype.hasOwnProperty.call(args, key)
                    ? args[key]
                    : undefined,
            });
        }, {});
    }
    injectFields(handler, instance, args, nativeOption) {
        this.fieldOptions(handler).forEach(data => {
            const key = `--${data.name}`;
            const hasValue = Object.prototype.hasOwnProperty.call(args, key);
            const value = hasValue ? args[key] : instance[data.propertyKey];
            if (data.required && value === undefined) {
                throw (0, errors_1.createRuntimePrintError)(errors_1.F_RUNTIME_PRINT.VALIDATION, errors_1.errorTypes.INPUT, `Option "--${data.name}" is required.`, { option: data.name });
            }
            instance[data.propertyKey] = value;
            nativeOption[data.name] = value;
        });
    }
    validateFieldValues(handler, nativeOption) {
        const validators = Reflect.getMetadata(metadata_1.metadata.VALUE_VALIDATOR_IDENTIFIER, handler) || {};
        this.fieldOptions(handler).forEach(data => {
            const fns = validators[data.propertyKey] || [];
            fns.forEach(fn => {
                try {
                    const result = fn(nativeOption[data.name], nativeOption);
                    if (result === false || typeof result === 'string') {
                        throw (0, errors_1.createRuntimePrintError)(errors_1.F_RUNTIME_PRINT.VALIDATION, errors_1.errorTypes.INPUT, typeof result === 'string'
                            ? result
                            : `Option "--${data.name}" is invalid.`, { option: data.name, value: nativeOption[data.name] });
                    }
                }
                catch (error) {
                    if ((0, errors_1.isFuncError)(error))
                        throw error;
                    throw (0, errors_1.createRuntimePrintError)(errors_1.F_RUNTIME_PRINT.VALIDATION, errors_1.errorTypes.INPUT, error instanceof Error ? error.message : String(error), { option: data.name, value: nativeOption[data.name] }, error instanceof Error ? error : undefined);
                }
            });
        });
    }
    selectHandler(handler, args, inputs) {
        const methodHandlers = this.methodHandlers(handler);
        const pathHandler = this.selectPathHandler(methodHandlers, inputs);
        if (pathHandler)
            return pathHandler;
        const matchedHandlers = methodHandlers.filter(data => {
            if (!data.flag)
                return false;
            return Object.prototype.hasOwnProperty.call(args, `--${data.flag}`);
        });
        if (matchedHandlers.length > 1) {
            throw (0, errors_1.createRuntimePrintError)(errors_1.F_RUNTIME_PRINT.MULTIPLE_OPTIONS, errors_1.errorTypes.INPUT, `Only one handler flag can be invoked at a time: ${matchedHandlers
                .map(item => `--${item.flag}`)
                .join(', ')}.`, { options: matchedHandlers.map(item => item.flag) });
        }
        const selectedHandler = matchedHandlers[0] || methodHandlers.find(data => !data.flag);
        if (selectedHandler) {
            return {
                data: selectedHandler,
                inputs,
                path: [],
            };
        }
        throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.MISSING_HANDLER, errors_1.errorTypes.REGISTRATION, `Command "${handler.name}" does not have a matching @Handler method.`, { handler: handler.name });
    }
    optionSpec(handler) {
        const fieldOptions = this.fieldOptions(handler);
        const handlerOptions = this.methodHandlers(handler)
            .filter(item => item.flag)
            .map(item => ({
            name: item.flag || '',
            alias: item.alias,
            description: item.description,
            type: Boolean,
        }));
        const options = [
            ...fieldOptions,
            ...handlerOptions,
            ...this.subOptions(handler),
        ];
        return filter.optionsToKeyValue(options);
    }
    fieldOptions(handler) {
        return Reflect.getMetadata(metadata_1.metadata.FIELD_OPTION_IDENTIFIER, handler) || [];
    }
    commandData(handler) {
        return Reflect.getMetadata(metadata_1.metadata.COMMAND_IDENTIFIER, handler);
    }
    methodHandlers(handler) {
        return Reflect.getMetadata(metadata_1.metadata.METHOD_HANDLER_IDENTIFIER, handler) || [];
    }
    selectPathHandler(handlers, inputs) {
        const matched = handlers
            .filter(data => data.path && data.path.length)
            .filter(data => data.path.every((item, index) => inputs[index] === item))
            .sort((a, b) => b.path.length - a.path.length)[0];
        if (!matched)
            return undefined;
        return {
            data: matched,
            inputs: inputs.slice(matched.path.length),
            path: matched.path || [],
        };
    }
    subOptions(handler) {
        return Reflect.getMetadata(metadata_1.metadata.SUB_OPTION_IDENTIFIER, handler) || [];
    }
    findCommand() {
        const first = this.argv[0];
        if (!first || first.startsWith('-'))
            return undefined;
        return this.registry.findCommand(first);
    }
    throwUnknownOptions(args, hasCommand) {
        const options = args._.filter((input, index) => {
            if (hasCommand && index === 0)
                return false;
            return input.startsWith('-') && input !== '-';
        });
        if (!options.length)
            return;
        throw (0, errors_1.createRuntimePrintError)(errors_1.F_RUNTIME_PRINT.UNKNOWN_OPTION, errors_1.errorTypes.INPUT, `Unknown option${options.length > 1 ? 's' : ''}: ${options.join(', ')}.`, { options });
    }
    dispatchError(error) {
        const funcError = (0, errors_1.normalizeRuntimeError)(error);
        if (funcError.level === errors_1.errorLevels.SYSTEM) {
            (0, errors_1.handleSystemError)(funcError);
        }
        const errors = this.registry.errors();
        if (!errors.length) {
            if (funcError.level === errors_1.errorLevels.RUNTIME_PRINT) {
                (0, errors_1.handleRuntimePrintError)(funcError, false);
                return;
            }
            throw funcError;
        }
        const factory = new factory_1.Factory({
            nativeOption: {},
            args: { _: [] },
            commands: this.registry.commands(),
            services: this.services,
        });
        const exception = new context_1.FuncException(funcError);
        errors.forEach(errorHandler => {
            const params = factory.getServiceParams(errorHandler, exception);
            new errorHandler(...params);
        });
        if (funcError.level === errors_1.errorLevels.RUNTIME_PRINT) {
            (0, errors_1.handleRuntimePrintError)(funcError, exception.printPrevented);
        }
    }
    async dispatchLocalCatch(handler, instance, error, factory) {
        const funcError = (0, errors_1.normalizeRuntimeError)(error);
        if (funcError.level === errors_1.errorLevels.SYSTEM)
            return false;
        const catches = Reflect.getMetadata(metadata_1.metadata.METHOD_CATCH_IDENTIFIER, handler) || [];
        if (!catches.length)
            return false;
        await catches.reduce(async (promise, data) => {
            await promise;
            const params = factory.getMethodServiceParams(instance, data.methodName, new context_1.FuncException(funcError));
            await Promise.resolve(instance[data.methodName](...params));
        }, Promise.resolve());
        return true;
    }
}
exports.Container = Container;
