// func/dist/containers/registration-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegistrationValidator = void 0;
const errors_1 = require("../errors");
const metadata_1 = require("../utils/metadata");
class RegistrationValidator {
    registry;
    constructor(registry) {
        this.registry = registry;
    }
    validate() {
        this.validateKnownHandlers();
        this.validateDuplicateTokens(this.registry.commands(), metadata_1.metadata.COMMAND_IDENTIFIER, data => [
            { token: data.name, type: errors_1.errorTokenTypes.COMMAND_NAME },
            { token: data.alias, type: errors_1.errorTokenTypes.COMMAND_ALIAS },
        ], errors_1.errorScopes.COMMAND);
        this.validateMajorCount();
        this.validateMissingCount();
        this.registry
            .commands()
            .concat(this.registry.majors())
            .concat(this.registry
            .missing()
            .filter(handler => this.methodHandlers(handler).length))
            .forEach(handler => this.validateCommandShape(handler));
    }
    validateKnownHandlers() {
        if (!this.registry.invalidHandlers.length)
            return;
        const names = this.registry.invalidHandlers.map(handler => handler.name || '<anonymous>');
        throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.UNKNOWN_HANDLER, errors_1.errorTypes.REGISTRATION, `All handlers passed to Container must use a func decorator: ${names.join(', ')}.`, { handlers: names });
    }
    validateMajorCount() {
        const majors = this.registry.majors();
        if (majors.length <= 1)
            return;
        throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.DUPLICATE_HANDLER, errors_1.errorTypes.REGISTRATION, `Only one major command can be registered: ${majors.map(item => item.name).join(', ')}.`, { scope: errors_1.errorScopes.COMMAND, handlers: majors.map(item => item.name) });
    }
    validateMissingCount() {
        const missing = this.registry.missing();
        if (missing.length <= 1)
            return;
        throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.DUPLICATE_HANDLER, errors_1.errorTypes.REGISTRATION, `Only one missing command can be registered: ${missing.map(item => item.name).join(', ')}.`, { scope: errors_1.errorScopes.COMMAND, handlers: missing.map(item => item.name) });
    }
    validateDuplicateTokens(handlers, ident, tokensFor, scope) {
        const tokens = {};
        handlers.forEach(handler => {
            const data = Reflect.getMetadata(ident, handler);
            tokensFor(data).forEach(item => {
                if (!item.token)
                    return;
                if (tokens[item.token]) {
                    throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.DUPLICATE_HANDLER, errors_1.errorTypes.REGISTRATION, `Duplicate ${scope} token "${item.token}" found in "${tokens[item.token]}" and "${handler.name}".`, {
                        scope,
                        token: item.token,
                        type: item.type,
                        handlers: [tokens[item.token], handler.name],
                    });
                }
                tokens[item.token] = handler.name;
            });
        });
    }
    validateCommandShape(handler) {
        const fieldOptions = Reflect.getMetadata(metadata_1.metadata.FIELD_OPTION_IDENTIFIER, handler) || [];
        const subOptions = Reflect.getMetadata(metadata_1.metadata.SUB_OPTION_IDENTIFIER, handler) || [];
        const methodHandlers = Reflect.getMetadata(metadata_1.metadata.METHOD_HANDLER_IDENTIFIER, handler) || [];
        if (!methodHandlers.length) {
            throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.MISSING_HANDLER, errors_1.errorTypes.REGISTRATION, `Command "${handler.name}" must define at least one @Handler method.`, { handler: handler.name });
        }
        let defaultHandler = '';
        const tokens = {};
        const registerToken = (token, owner, type) => {
            if (!token)
                return;
            if (tokens[token]) {
                throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.DUPLICATE_HANDLER, errors_1.errorTypes.REGISTRATION, `Duplicate option token "${token}" found in "${handler.name}".`, {
                    handler: handler.name,
                    owner,
                    scope: errors_1.errorScopes.COMMAND,
                    token,
                    type,
                });
            }
            tokens[token] = owner;
        };
        fieldOptions.forEach(data => {
            this.validateOptionType(data, handler.name);
            registerToken(`--${data.name}`, data.propertyKey, errors_1.errorTokenTypes.OPTION_NAME);
            registerToken(data.alias ? `-${data.alias}` : undefined, data.propertyKey, errors_1.errorTokenTypes.OPTION_ALIAS);
        });
        subOptions.forEach(data => {
            this.validateOptionType(data, handler.name);
            registerToken(`--${data.name}`, `SubOptions.${data.name}`, errors_1.errorTokenTypes.OPTION_NAME);
            registerToken(data.alias ? `-${data.alias}` : undefined, `SubOptions.${data.name}`, errors_1.errorTokenTypes.OPTION_ALIAS);
        });
        methodHandlers.forEach(data => {
            if (data.flag && data.path) {
                throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.INVALID_PARAM_VALUE, errors_1.errorTypes.REGISTRATION, `Handler "${data.methodName}" in "${handler.name}" cannot define both flag and path.`, { handler: handler.name, method: data.methodName });
            }
            if (!data.flag) {
                if (data.path && data.path.length) {
                    registerToken(data.path.join(' '), data.methodName, errors_1.errorTokenTypes.COMMAND_NAME);
                    return;
                }
                if (defaultHandler) {
                    throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.DUPLICATE_HANDLER, errors_1.errorTypes.REGISTRATION, `Command "${handler.name}" must define only one default @Handler method.`, { handler: handler.name, methods: [defaultHandler, data.methodName] });
                }
                defaultHandler = data.methodName;
                return;
            }
            registerToken(`--${data.flag}`, data.methodName, errors_1.errorTokenTypes.OPTION_NAME);
            registerToken(data.alias ? `-${data.alias}` : undefined, data.methodName, errors_1.errorTokenTypes.OPTION_ALIAS);
        });
    }
    validateOptionType(data, handlerName) {
        if (data.type !== Array)
            return;
        throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.UNSUPPORTED_ARRAY_TYPE, errors_1.errorTypes.REGISTRATION, `Option "${data.name}" in "${handlerName}" uses Array. Please use [String] instead.`, { option: data.name, handler: handlerName });
    }
    methodHandlers(handler) {
        return Reflect.getMetadata(metadata_1.metadata.METHOD_HANDLER_IDENTIFIER, handler) || [];
    }
}
exports.RegistrationValidator = RegistrationValidator;
