// func/dist/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommandRegistry = exports.FuncException = void 0;
class FuncException {
    _error;
    defaultPrintPrevented = false;
    constructor(_error) {
        this._error = _error;
    }
    get code() {
        return this._error.code;
    }
    get details() {
        return this._error.details;
    }
    get error() {
        return this._error;
    }
    get level() {
        return this._error.level;
    }
    get message() {
        return this._error.message;
    }
    get printPrevented() {
        return this.defaultPrintPrevented;
    }
    get type() {
        return this._error.type;
    }
    preventDefaultPrint() {
        this.defaultPrintPrevented = true;
    }
}
exports.FuncException = FuncException;
class CommandRegistry {
    _commands;
    constructor(_commands = []) {
        this._commands = _commands;
    }
    get commands() {
        return this._commands;
    }
}
exports.CommandRegistry = CommandRegistry;
