// func/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Container = exports.run = exports.createApp = exports.Service = exports.FuncModule = exports.ValueValidate = exports.Value = exports.Required = exports.Flag = exports.Exclusive = exports.Enum = exports.DependsOn = exports.ArrayValue = exports.Regs = exports.Exception = exports.Args = exports.SubOptions = exports.Handler = exports.CommandMissing = exports.CommandMajor = exports.CommandError = exports.Command = exports.CatchAll = exports.Catch = void 0;
require("reflect-metadata");
var commands_1 = require("./commands");
Object.defineProperty(exports, "Catch", { enumerable: true, get: function () { return commands_1.Catch; } });
Object.defineProperty(exports, "CatchAll", { enumerable: true, get: function () { return commands_1.CatchAll; } });
Object.defineProperty(exports, "Command", { enumerable: true, get: function () { return commands_1.Command; } });
Object.defineProperty(exports, "CommandError", { enumerable: true, get: function () { return commands_1.CommandError; } });
Object.defineProperty(exports, "CommandMajor", { enumerable: true, get: function () { return commands_1.CommandMajor; } });
Object.defineProperty(exports, "CommandMissing", { enumerable: true, get: function () { return commands_1.CommandMissing; } });
Object.defineProperty(exports, "Handler", { enumerable: true, get: function () { return commands_1.Handler; } });
Object.defineProperty(exports, "SubOptions", { enumerable: true, get: function () { return commands_1.SubOptions; } });
var injection_1 = require("./injection");
Object.defineProperty(exports, "Args", { enumerable: true, get: function () { return injection_1.Args; } });
Object.defineProperty(exports, "Exception", { enumerable: true, get: function () { return injection_1.Exception; } });
Object.defineProperty(exports, "Regs", { enumerable: true, get: function () { return injection_1.Regs; } });
var options_1 = require("./options");
Object.defineProperty(exports, "ArrayValue", { enumerable: true, get: function () { return options_1.ArrayValue; } });
Object.defineProperty(exports, "DependsOn", { enumerable: true, get: function () { return options_1.DependsOn; } });
Object.defineProperty(exports, "Enum", { enumerable: true, get: function () { return options_1.Enum; } });
Object.defineProperty(exports, "Exclusive", { enumerable: true, get: function () { return options_1.Exclusive; } });
Object.defineProperty(exports, "Flag", { enumerable: true, get: function () { return options_1.Flag; } });
Object.defineProperty(exports, "Required", { enumerable: true, get: function () { return options_1.Required; } });
Object.defineProperty(exports, "Value", { enumerable: true, get: function () { return options_1.Value; } });
Object.defineProperty(exports, "ValueValidate", { enumerable: true, get: function () { return options_1.ValueValidate; } });
var application_1 = require("./application");
Object.defineProperty(exports, "FuncModule", { enumerable: true, get: function () { return application_1.FuncModule; } });
Object.defineProperty(exports, "Service", { enumerable: true, get: function () { return application_1.Service; } });
Object.defineProperty(exports, "createApp", { enumerable: true, get: function () { return application_1.createApp; } });
Object.defineProperty(exports, "run", { enumerable: true, get: function () { return application_1.run; } });
var container_1 = require("./containers/container");
Object.defineProperty(exports, "Container", { enumerable: true, get: function () { return container_1.Container; } });
__exportStar(require("./errors"), exports);
__exportStar(require("./context"), exports);
