// func/dist/injection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Regs = exports.Exception = exports.Args = exports.injectTokens = void 0;
require("reflect-metadata");
const metadata_1 = require("./utils/metadata");
var injectTokens;
(function (injectTokens) {
    injectTokens["ARGS"] = "args";
    injectTokens["EXCEPTION"] = "exception";
    injectTokens["REGS"] = "regs";
})(injectTokens || (exports.injectTokens = injectTokens = {}));
const injectToken = (token) => (target, propertyKey, parameterIndex) => {
    const metadataTarget = propertyKey ? target : target;
    const tokens = propertyKey
        ? Reflect.getMetadata(metadata_1.metadata.PARAM_INJECT_TOKEN_IDENTIFIER, metadataTarget, propertyKey) || {}
        : Reflect.getMetadata(metadata_1.metadata.PARAM_INJECT_TOKEN_IDENTIFIER, metadataTarget) || {};
    if (propertyKey) {
        Reflect.defineMetadata(metadata_1.metadata.PARAM_INJECT_TOKEN_IDENTIFIER, Object.assign({}, tokens, { [parameterIndex]: token }), metadataTarget, propertyKey);
        return;
    }
    Reflect.defineMetadata(metadata_1.metadata.PARAM_INJECT_TOKEN_IDENTIFIER, Object.assign({}, tokens, { [parameterIndex]: token }), metadataTarget);
};
const Args = () => injectToken(injectTokens.ARGS);
exports.Args = Args;
const Exception = () => injectToken(injectTokens.EXCEPTION);
exports.Exception = Exception;
const Regs = () => injectToken(injectTokens.REGS);
exports.Regs = Regs;
