// func/dist/options/constraint-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConstraintValidator = void 0;
const errors_1 = require("../errors");
class ConstraintValidator {
    fieldOptions;
    args;
    option;
    constructor(fieldOptions, args, option) {
        this.fieldOptions = fieldOptions;
        this.args = args;
        this.option = option;
    }
    validate() {
        this.fieldOptions.forEach(data => {
            this.validateEnum(data);
            this.validateDependsOn(data);
            this.validateExclusive(data);
        });
    }
    explicitNames() {
        return new Set(this.fieldOptions
            .filter(data => Object.prototype.hasOwnProperty.call(this.args, `--${data.name}`))
            .map(data => data.name));
    }
    validateDependsOn(data) {
        if (!data.dependsOn || !data.dependsOn.length)
            return;
        const explicit = this.explicitNames();
        if (!explicit.has(data.name))
            return;
        const missing = data.dependsOn.filter(item => !explicit.has(item));
        if (!missing.length)
            return;
        throw (0, errors_1.createRuntimePrintError)(errors_1.F_RUNTIME_PRINT.VALIDATION, errors_1.errorTypes.INPUT, `Option "--${data.name}" depends on: ${missing.map(item => `--${item}`).join(', ')}.`, { option: data.name, dependsOn: data.dependsOn, missing });
    }
    validateEnum(data) {
        if (!data.enum || !data.enum.length)
            return;
        const value = this.option[data.name];
        if (value === undefined)
            return;
        const values = Array.isArray(value) ? value : [value];
        const invalid = values.filter(item => !data.enum.includes(item));
        if (!invalid.length)
            return;
        throw (0, errors_1.createRuntimePrintError)(errors_1.F_RUNTIME_PRINT.VALIDATION, errors_1.errorTypes.INPUT, `Option "--${data.name}" must be one of: ${data.enum.join(', ')}.`, { option: data.name, enum: data.enum, invalid });
    }
    validateExclusive(data) {
        if (!data.exclusive || !data.exclusive.length)
            return;
        const explicit = this.explicitNames();
        if (!explicit.has(data.name))
            return;
        const conflicts = data.exclusive.filter(item => explicit.has(item));
        if (!conflicts.length)
            return;
        throw (0, errors_1.createRuntimePrintError)(errors_1.F_RUNTIME_PRINT.VALIDATION, errors_1.errorTypes.INPUT, `Option "--${data.name}" cannot be used with: ${conflicts.map(item => `--${item}`).join(', ')}.`, { option: data.name, exclusive: data.exclusive, conflicts });
    }
}
exports.ConstraintValidator = ConstraintValidator;
