// func/dist/options/field-options.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArrayValue = exports.Value = exports.Flag = void 0;
require("reflect-metadata");
const errors_1 = require("../errors");
const validator = __importStar(require("../utils/validator"));
const metadata_1 = require("../utils/metadata");
const fieldOptionFactory = (kind, params = {}) => (target, propertyKey) => {
    if (typeof target === 'function' || typeof propertyKey !== 'string') {
        throwInvalidFieldTarget(propertyKey);
    }
    const key = propertyKey;
    const name = Object.prototype.hasOwnProperty.call(params, 'name') ? params.name || '' : key;
    validator.optionName(name, 'name');
    validator.optionAlias(params.alias, 'alias');
    const type = resolveType(kind, params, target, key);
    const requiredKeys = Reflect.getMetadata(metadata_1.metadata.REQUIRED_FIELD_IDENTIFIER, target.constructor) || [];
    const constraints = Reflect.getMetadata(metadata_1.metadata.FIELD_CONSTRAINT_IDENTIFIER, target.constructor) || {};
    const nextOption = Object.assign({}, {
        kind,
        name,
        propertyKey: key,
        required: requiredKeys.includes(key),
        type,
    }, params, constraints[key] || {});
    const options = Reflect.getMetadata(metadata_1.metadata.FIELD_OPTION_IDENTIFIER, target.constructor) || [];
    Reflect.defineMetadata(metadata_1.metadata.FIELD_OPTION_IDENTIFIER, options.concat([nextOption]), target.constructor);
};
const resolveType = (kind, params, target, propertyKey) => {
    if (kind === 'flag')
        return Boolean;
    if (kind === 'array')
        return [String];
    const explicitType = params.type;
    if (explicitType)
        return explicitType;
    const designType = Reflect.getMetadata(metadata_1.metadata.DESIGN_TYPE, target, propertyKey);
    if (designType === String || designType === Number || designType === Boolean)
        return designType;
    throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.INVALID_PARAM_TYPE, errors_1.errorTypes.DEFINITION, `Cannot infer value type for "${propertyKey}". Please pass type explicitly.`, {
        property: propertyKey,
        reason: 'cannot-infer-value-type',
    });
};
const throwInvalidFieldTarget = (propertyKey) => {
    throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.INVALID_PARAM_TYPE, errors_1.errorTypes.DEFINITION, `Field option "${String(propertyKey)}" must decorate an instance property.`, { property: String(propertyKey) });
};
const Flag = (params = {}) => fieldOptionFactory('flag', params);
exports.Flag = Flag;
const Value = (params = {}) => fieldOptionFactory('value', params);
exports.Value = Value;
const ArrayValue = (params = {}) => fieldOptionFactory('array', params);
exports.ArrayValue = ArrayValue;
