// func/dist/utils/factory.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Factory = void 0;
const filter = __importStar(require("./filter"));
const metadata_1 = require("./metadata");
const injector_1 = require("../application/injector");
const injection_1 = require("../injection");
const context_1 = require("../context");
class Factory {
    params;
    injector;
    constructor(params) {
        this.params = params;
        this.injector = new injector_1.ServiceInjector(params.services);
    }
    getServiceParams(target, error) {
        return this.resolveServiceParams(target, undefined, error, []);
    }
    getMethodServiceParams(instance, methodName, error) {
        const target = Object.getPrototypeOf(instance);
        return this.resolveServiceParams(target, methodName, error, []);
    }
    resolveServiceParams(target, propertyKey, error, serviceStack = []) {
        const injectedTokens = this.injectedTokens(target, propertyKey);
        const paramTypes = this.paramTypes(target, propertyKey);
        const args = this.params.args;
        const option = this.params.nativeOption;
        const inputs = this.params.inputs || args._;
        const length = Math.max(paramTypes.length, 0, ...Object.keys(injectedTokens).map(item => Number(item) + 1));
        return Array.from({ length }).map((_, index) => {
            const token = injectedTokens[index];
            if (token)
                return this.resolveInjectedToken(token, inputs, option, args, error);
            return this.resolveService(paramTypes[index], error, serviceStack);
        });
    }
    injectedTokens(target, propertyKey) {
        if (propertyKey) {
            return Reflect.getMetadata(metadata_1.metadata.PARAM_INJECT_TOKEN_IDENTIFIER, target, propertyKey) || {};
        }
        return Reflect.getMetadata(metadata_1.metadata.PARAM_INJECT_TOKEN_IDENTIFIER, target) || {};
    }
    paramTypes(target, propertyKey) {
        if (propertyKey) {
            return Reflect.getMetadata(metadata_1.metadata.DESIGN_PARAM_TYPES, target, propertyKey) || [];
        }
        return Reflect.getMetadata(metadata_1.metadata.DESIGN_PARAM_TYPES, target) || [];
    }
    resolveService(target, error, serviceStack = []) {
        return this.injector.resolve(target, (service, nextStack) => this.resolveServiceParams(service, undefined, error, nextStack), serviceStack);
    }
    resolveInjectedToken(token, inputs, option, args, error) {
        if (token === injection_1.injectTokens.ARGS) {
            return {
                command: this.params.command,
                handler: this.params.handler,
                inputs,
                native: args,
                option,
                path: this.params.path || [],
            };
        }
        if (token === injection_1.injectTokens.EXCEPTION) {
            if (error instanceof context_1.FuncException)
                return error;
            return new context_1.FuncException(error);
        }
        if (token === injection_1.injectTokens.REGS) {
            const commands = filter.commandsToDatas(this.params.commands);
            return new context_1.CommandRegistry(commands);
        }
        return undefined;
    }
}
exports.Factory = Factory;
