// func/dist/utils/validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mustBeArray = exports.optionAlias = exports.optionName = exports.commandAlias = exports.commandName = exports.requireKey = void 0;
const errors_1 = require("../errors");
const requireKey = (value, key) => {
    if (!value) {
        throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.MISSING_REQUIRED_PARAM, errors_1.errorTypes.DEFINITION, `Param \`${key}\` is required.`, { key });
    }
};
exports.requireKey = requireKey;
const commandName = (value, key) => {
    (0, exports.requireKey)(value, key);
    token(value, key);
};
exports.commandName = commandName;
const commandAlias = (value, key) => {
    if (!value)
        return;
    token(value, key);
};
exports.commandAlias = commandAlias;
const optionName = (value, key) => {
    (0, exports.requireKey)(value, key);
    token(value, key);
};
exports.optionName = optionName;
const optionAlias = (value, key) => {
    if (!value)
        return;
    token(value, key);
    if (value.length === 1)
        return;
    throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.INVALID_PARAM_VALUE, errors_1.errorTypes.DEFINITION, `Param \`${key}\` must be a single character.`, { key, value });
};
exports.optionAlias = optionAlias;
const mustBeArray = (value, key) => {
    if (!Array.isArray(value)) {
        throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.INVALID_PARAM_TYPE, errors_1.errorTypes.DEFINITION, `Param \`${key}\` must be \`Array\`.`, { key, expected: 'Array' });
    }
};
exports.mustBeArray = mustBeArray;
const token = (value, key) => {
    if (value.trim() === value && !value.startsWith('-') && !/\s/.test(value))
        return;
    throw (0, errors_1.createSystemError)(errors_1.F_SYSTEM.INVALID_PARAM_VALUE, errors_1.errorTypes.DEFINITION, `Param \`${key}\` cannot be empty, start with "-", or contain whitespace.`, { key, value });
};
