// func/dist/src/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleEffect = exports.handleRuntimePrintError = exports.handleSystemError = exports.normalizeRuntimeError = exports.createRuntimePrintError = exports.createRuntimeError = exports.createSystemError = exports.isFuncError = exports.FuncError = exports.errorTokenTypes = exports.errorScopes = exports.errorTypes = exports.errorLevels = exports.F_EFFECT = exports.F_RUNTIME_PRINT = exports.F_RUNTIME = exports.F_SYSTEM = void 0;
exports.F_SYSTEM = {
    DUPLICATE_HANDLER: 'F_SYSTEM_DUPLICATE_HANDLER',
    INVALID_PARAM_TYPE: 'F_SYSTEM_INVALID_PARAM_TYPE',
    INVALID_PARAM_VALUE: 'F_SYSTEM_INVALID_PARAM_VALUE',
    MISSING_PROVIDER: 'F_SYSTEM_MISSING_PROVIDER',
    MISSING_HANDLER: 'F_SYSTEM_MISSING_HANDLER',
    MISSING_PARAM_TYPES: 'F_SYSTEM_MISSING_PARAM_TYPES',
    MISSING_REQUIRED_PARAM: 'F_SYSTEM_MISSING_REQUIRED_PARAM',
    UNKNOWN_HANDLER: 'F_SYSTEM_UNKNOWN_HANDLER',
    UNSUPPORTED_ARRAY_TYPE: 'F_SYSTEM_UNSUPPORTED_ARRAY_TYPE',
};
exports.F_RUNTIME = {
    HANDLER_ERROR: 'F_RUNTIME_HANDLER_ERROR',
};
exports.F_RUNTIME_PRINT = {
    MULTIPLE_OPTIONS: 'F_RUNTIME_PRINT_MULTIPLE_OPTIONS',
    PARSE: 'F_RUNTIME_PRINT_PARSE',
    UNKNOWN_OPTION: 'F_RUNTIME_PRINT_UNKNOWN_OPTION',
    VALIDATION: 'F_RUNTIME_PRINT_VALIDATION',
};
exports.F_EFFECT = {
    DEPRECATED_API: 'F_EFFECT_DEPRECATED_API',
};
var errorLevels;
(function (errorLevels) {
    errorLevels["EFFECT"] = "effect";
    errorLevels["RUNTIME"] = "runtime";
    errorLevels["RUNTIME_PRINT"] = "runtime-print";
    errorLevels["SYSTEM"] = "system";
})(errorLevels || (exports.errorLevels = errorLevels = {}));
var errorTypes;
(function (errorTypes) {
    errorTypes["DEFINITION"] = "definition";
    errorTypes["DEPRECATION"] = "deprecation";
    errorTypes["HANDLER"] = "handler";
    errorTypes["INJECTION"] = "injection";
    errorTypes["INPUT"] = "input";
    errorTypes["REGISTRATION"] = "registration";
})(errorTypes || (exports.errorTypes = errorTypes = {}));
var errorScopes;
(function (errorScopes) {
    errorScopes["COMMAND"] = "command";
})(errorScopes || (exports.errorScopes = errorScopes = {}));
var errorTokenTypes;
(function (errorTokenTypes) {
    errorTokenTypes["COMMAND_ALIAS"] = "command alias";
    errorTokenTypes["COMMAND_NAME"] = "command name";
    errorTokenTypes["OPTION_ALIAS"] = "option alias";
    errorTokenTypes["OPTION_NAME"] = "option name";
})(errorTokenTypes || (exports.errorTokenTypes = errorTokenTypes = {}));
class FuncError extends Error {
    cause;
    code;
    details;
    level;
    type;
    constructor(params) {
        super(params.message);
        this.name = 'FuncError';
        this.code = params.code;
        this.details = params.details || {};
        this.level = params.level;
        this.type = params.type;
        this.cause = params.cause;
    }
}
exports.FuncError = FuncError;
const isFuncError = (error) => error instanceof FuncError;
exports.isFuncError = isFuncError;
const createSystemError = (code, type, message, details = {}, cause) => new FuncError({
    code,
    details,
    level: errorLevels.SYSTEM,
    message,
    type,
    cause,
});
exports.createSystemError = createSystemError;
const createRuntimeError = (code, type, message, details = {}, cause) => new FuncError({
    code,
    details,
    level: errorLevels.RUNTIME,
    message,
    type,
    cause,
});
exports.createRuntimeError = createRuntimeError;
const createRuntimePrintError = (code, type, message, details = {}, cause) => new FuncError({
    code,
    details,
    level: errorLevels.RUNTIME_PRINT,
    message,
    type,
    cause,
});
exports.createRuntimePrintError = createRuntimePrintError;
const normalizeRuntimeError = (error) => {
    if ((0, exports.isFuncError)(error))
        return error;
    const nativeError = error instanceof Error ? error : new Error(String(error));
    return (0, exports.createRuntimeError)(exports.F_RUNTIME.HANDLER_ERROR, errorTypes.HANDLER, nativeError.message, { error: nativeError }, nativeError);
};
exports.normalizeRuntimeError = normalizeRuntimeError;
const handleSystemError = (error) => {
    throw error;
};
exports.handleSystemError = handleSystemError;
const handleRuntimePrintError = (error, printPrevented) => {
    if (printPrevented)
        return;
    console.error(error.message);
};
exports.handleRuntimePrintError = handleRuntimePrintError;
const handleEffect = (code, type, message, details = {}) => {
    const error = new FuncError({
        code,
        details,
        level: errorLevels.EFFECT,
        message,
        type,
    });
    console.warn(error.message);
    return error;
};
exports.handleEffect = handleEffect;
