// func/dist/src/containers/handler-registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HandlerRegistry = void 0;
const metadata_1 = require("../utils/metadata");
class HandlerRegistry {
    datas = {};
    invalidHandlers = [];
    constructor(params) {
        params.forEach(handler => {
            const type = Reflect.getMetadata(metadata_1.metadata.HANDLER_IDENTIFIER, handler);
            if (!type) {
                this.invalidHandlers = this.invalidHandlers.concat([handler]);
                return;
            }
            this.datas[type] = (this.datas[type] || []).concat([handler]);
        });
    }
    commands() {
        return this.handlers(metadata_1.handlers.COMMAND);
    }
    errors() {
        return this.handlers(metadata_1.handlers.ERROR);
    }
    major() {
        return this.majors()[0];
    }
    majors() {
        return this.handlers(metadata_1.handlers.MAJOR);
    }
    missing() {
        return this.handlers(metadata_1.handlers.MISSING);
    }
    findCommand(input) {
        if (!input || input.startsWith('-'))
            return undefined;
        return this.commands().find(item => {
            const data = Reflect.getMetadata(metadata_1.metadata.COMMAND_IDENTIFIER, item);
            if (!data)
                return false;
            return data.name === input || data.alias === input;
        });
    }
    handlers(kind) {
        return this.datas[kind] || [];
    }
}
exports.HandlerRegistry = HandlerRegistry;
